lib/ide/src/controller.rs

```diff
@@ -0,0 +1,426 @@
+//! This module contains all the controllers. They cover everything that is
+//! between clients of remote services (like language server and file manager)
+//! and views.
+//!
+//! The controllers create a tree-like structure, with project controller being
+//! a root, then module controllers below, then graph/text controller and so on.
+//!
+//! As a general rule, while the "upper" (i.e. closer to root) nodes may keep
+//! handles to the "lower" nodes (e.g. to allow their reuse), they should never
+//! manage their lifetime.
+//!
+//! Primarily views are considered owners of their respective controllers.
+//! Additionally, controllers are allowed to keep strong handle "upwards".
+//!
+//! Controllers store their handles using `utils::cell` handle types to ensure
+//! that mutable state is safely accessed.
+
+use crate::prelude::*;
+
+/// General-purpose `Result` supporting any `Error`-compatible failures.
+pub type FallibleResult<T> = Result<T,failure::Error>;
+
+/// Macro defines `StrongHandle` and `WeakHandle` newtypes for handles storing
+/// the type given in the argument.
+///
+/// This allows treating handles as separate types and fitting them with impl
+/// methods of their own. While not necessary, such implementation may allow
+/// hiding from user gritty details of `with` usage behind nice, easy API.
+macro_rules! make_handles {
+    ($data_type:ty) => {
+        /// newtype wrapper over StrongHandle.
+        #[derive(Shrinkwrap)]
+        #[derive(Clone,Debug)]
+        pub struct StrongHandle(pub utils::cell::StrongHandle<$data_type>);
+
+        impl StrongHandle {
+            /// Obtain a WeakHandle to this data.
+            pub fn downgrade(&self) -> WeakHandle {
+                WeakHandle(self.0.downgrade())
+            }
+            /// Create a new StrongHandle that will wrap given data.
+            pub fn new(data:$data_type) -> StrongHandle {
+                StrongHandle(utils::cell::StrongHandle::new(data))
+            }
+        }
+
+        /// newtype wrapper over WeakHandle.
+        #[derive(Shrinkwrap)]
+        #[derive(Clone,Debug)]
+        pub struct WeakHandle(pub utils::cell::WeakHandle<$data_type>);
+
+        impl WeakHandle {
+            /// Obtain a StrongHandle to this data.
+            pub fn upgrade(&self) -> Option<StrongHandle> {
+                self.0.upgrade().map(StrongHandle)
+            }
+        }
+    };
+}
+
+
+
+// =======================
+// === Text controller ===
+// =======================
+
+pub mod project {
+    //! Project controller.
```

Comment:
What does "project" mean here? Is this really a "project" - a set of Luna source files in a particular location on disk? If not, the name should be changed. Otherwise we will use "prject" for two different things.

Reply:
this is module for project controller. It would be a little silly to write "controller::project_controller" wouldn't it?